crates/valence_client/src/lib.rs

```diff
@@ -934,13 +960,33 @@ fn read_data_in_old_view(
                     for &(id, dest_pos) in &cell.outgoing {
                         if dest_pos.map_or(true, |p| !view.contains(p)) {
                             // The outgoing entity moved outside the view distance, so it must be
-                            // despawned.
-                            if let Ok(entity_id) = entity_ids.get(id) {
-                                remove_buf.push(entity_id.get());
+                            // despawned if it was in the client's layer mask.
+                            if let Ok((entity_id, layer)) = entity_ids.get(id) {
+                                if let Some(client_layer_set) = client_layer_set {
+                                    if let Some(layer) = layer {
+                                        if client_layer_set.contains(&layer.0) {
+                                            remove_buf.push(entity_id.get());
+                                        }
+                                    } else {
+                                        remove_buf.push(entity_id.get());
+                                    }
+                                } else {
+                                    remove_buf.push(entity_id.get());
+                                }
```

Comment:
Nitpicking: group ifs `if let (Some(client_layer_set), Some(layer)) = (client_layer_set, layer)` so that we can remove the duplicate branch like you did earlier